pkg/controller/controller_binding.go

```diff
@@ -581,108 +589,220 @@ func (c *controller) reconcileServiceBinding(binding *v1beta1.ServiceBinding) er
 		return nil
 	}
 
-	// All updates not having a DeletingTimestamp will have been handled above
-	// and returned early, except in the case of orphan mitigation. Otherwise,
-	// when we reach this point, we're dealing with an update that's actually
-	// a soft delete-- i.e. we have some finalization to do.
-	if finalizers := sets.NewString(binding.Finalizers...); finalizers.Has(v1beta1.FinalizerServiceCatalog) || binding.Status.OrphanMitigationInProgress {
-		err := c.ejectServiceBinding(binding)
-		if err != nil {
-			s := fmt.Sprintf(`Error deleting secret: %s`, err)
-			glog.Warning(pcb.Message(s))
-			c.recorder.Eventf(binding, corev1.EventTypeWarning, errorEjectingBindReason, "%v %v", errorEjectingBindMessage, s)
-			setServiceBindingCondition(
-				toUpdate,
-				v1beta1.ServiceBindingConditionReady,
-				v1beta1.ConditionUnknown,
-				errorEjectingBindReason,
-				errorEjectingBindMessage+s,
-			)
-			if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
-				return err
-			}
+	return nil
+}
+
+func (c *controller) reconcileServiceBindingDelete(binding *v1beta1.ServiceBinding) error {
+	// All updates having a DeletingTimestamp will have been handled here.
+	// We're dealing with an update that's actually a soft delete-- i.e. we
+	// have some finalization to do.
+
+	if binding.DeletionTimestamp == nil && !binding.Status.OrphanMitigationInProgress {
+		// nothing to do...
+		return nil
+	}
+
+	pcb := pretty.NewContextBuilder(pretty.ServiceBinding, binding.Namespace, binding.Name)
+	glog.V(4).Info(pcb.Message("Processing Delete"))
+
+	finalizerToken := v1beta1.FinalizerServiceCatalog
+	finalizers := sets.NewString(binding.Finalizers...)
+	if !finalizers.Has(finalizerToken) {
+		return nil
+	}
+
+	// If unbind has failed, do not do anything more
+	if binding.Status.UnbindStatus == v1beta1.ServiceBindingUnbindStatusFailed {
+		glog.V(4).Info(pcb.Message("Not processing delete event because unbinding has failed"))
+		return nil
+	}
+
+	toUpdate, err := makeServiceBindingClone(binding)
+	if err != nil {
+		return err
+	}
+
+	// If unbinding succeeded or is not needed, then clear out the finalizers
+	if binding.Status.UnbindStatus == v1beta1.ServiceBindingUnbindStatusNotRequired ||
+		binding.Status.UnbindStatus == v1beta1.ServiceBindingUnbindStatusSucceeded {
+
+		glog.V(5).Info(pcb.Message("Clearing catalog finalizer"))
+
+		// Clear the finalizer
+		finalizers.Delete(v1beta1.FinalizerServiceCatalog)
+		toUpdate.Finalizers = finalizers.List()
+
+		toUpdate.Status.ExternalProperties = nil
```

Comment:
We should not need to clear out external properties or current operation here. If the `UnbindStatus` is `NotRequired` or `Succeeded`, then there must not be a pending current operation being performed on the binding.